### PR TITLE
Add facades for ES6 Iterator and Iterable

### PR DESCRIPTION
**Pull request by gzm0**
Also add conversions to and from relevant Scala types. Note that these facades do not take return values into account (which are needed for full generator typing).

### CODE REVIEW

**Comment by sjrd** [comment]
Yes, that looks reasonable. I think I had not made `js.Array` extend `js.Iterable` because collection operations on `js.Array` should use `WrappedArray`, but the lower priority implicit is better. 

Btw I think we have precedent for naming that one Low*est*ProrityImplicits.

**Comment by gzm0** [comment]
Updated. Review by @sjrd

File: library/src/main/scala/scala/scalajs/js/Iterable.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2015, LAMP/EPFL   **
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/Iterable.scala]: 2017.

Seems we should have some global pass over these header comments, one of these days.

File: library/src/main/scala/scala/scalajs/js/IterableOps.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-lang.org/     **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+package scala.scalajs.js
+
+/** Adapts a JavaScript Iterable to a Scala Iterable */
+@inline
+final class IterableOps[+A](private[this] val self: Iterable[A])
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/IterableOps.scala]: Remove `private[this] val`?

File: library/src/main/scala/scala/scalajs/js/Iterator.scala [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    @inline
+    def next(): A = {
+      val value = lastEntry.value
+      lastEntry = self.next()
+      value
+    }
+  }
+
+  @inline
+  final implicit class IteratorOps[A](private[this] val self: Iterator[A]) {
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/Iterator.scala]: This should be a value class, instead of an `@inline` reference class.

File: library/src/main/scala/scala/scalajs/js/JSConverters.scala [hunk 4]
```diff
@@ -41,6 +42,46 @@ object JSConverters extends JSConvertersLowPrioImplicits {
     @inline final def toJSArray: Array[T] = genTraversableOnce2jsArray(col)
   }
 
+  implicit class JSRichGenIterable[T](val col: GenIterable[T]) extends AnyVal {
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSConverters.scala]: Use obscure names like `__self` for the underlying field of an `implicit` value class, so that it's obvious it should not be used directly on a `GenIterable`.

File: library/src/main/scala/scala/scalajs/js/JSConverters.scala [hunk 5]
```diff
@@ -41,6 +42,46 @@ object JSConverters extends JSConvertersLowPrioImplicits {
...
+    @inline final def toJSIterable: Iterable[T] = new IterableAdapter(col)
+  }
+
+  implicit class JSRichIterator[T](
+      val it: scala.collection.Iterator[T]) extends AnyVal {
+    @inline final def toJSIterator: Iterator[T] = new IteratorAdapter(it)
+  }
+
+  @inline
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSConverters.scala]: `@inline` on a JS class doesn't do anything. I suggest you remove it so that we do not pretend it does.

File: library/src/main/scala/scala/scalajs/js/JSConverters.scala [hunk 6]
```diff
@@ -41,6 +42,46 @@ object JSConverters extends JSConvertersLowPrioImplicits {
...
+  @ScalaJSDefined
+  private class IterableAdapter[+T](private[this] val col: GenIterable[T])
+      extends Iterable[T] {
+    @inline
+    @JSName(Symbol.iterator)
+    final def jsIterator(): Iterator[T] = col.iterator.toJSIterator
+  }
+
+  @inline
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSConverters.scala]: Same here.

File: library/src/main/scala/scala/scalajs/js/JSConverters.scala [hunk 7]
```diff
@@ -41,6 +42,46 @@ object JSConverters extends JSConvertersLowPrioImplicits {
...
+  implicit class JSRichIterator[T](
+      val it: scala.collection.Iterator[T]) extends AnyVal {
+    @inline final def toJSIterator: Iterator[T] = new IteratorAdapter(it)
+  }
+
+  @inline
+  @ScalaJSDefined
+  private class IterableAdapter[+T](private[this] val col: GenIterable[T])
+      extends Iterable[T] {
+    @inline
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSConverters.scala]: Same here.

File: library/src/main/scala/scala/scalajs/js/JSConverters.scala [hunk 8]
```diff
@@ -41,6 +42,46 @@ object JSConverters extends JSConvertersLowPrioImplicits {
...
+    @JSName(Symbol.iterator)
+    final def jsIterator(): Iterator[T] = col.iterator.toJSIterator
+  }
+
+  @inline
+  @ScalaJSDefined
+  private class IteratorAdapter[+T](
+      private[this] val it: scala.collection.Iterator[T]) extends Iterator[T] {
+    @inline
```
Comment (sjrd) [inline library/src/main/scala/scala/scalajs/js/JSConverters.scala]: And same here.

File: test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/IterableTest.scala [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+import org.junit.Assume._
+import org.junit.{BeforeClass, Test}
+
+object IterableTest {
+  @BeforeClass def assumeSymbolsAreSupported(): Unit = {
+    assumeTrue("Assuming JavaScript symbols are supported",
+        org.scalajs.testsuite.utils.Platform.areJSSymbolsSupported)
+  }
+
+  @inline
```
Comment (sjrd) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/IterableTest.scala]: Same: `@inline` is not doing anything here.

File: test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/IterableTest.scala [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+  @ScalaJSDefined
+  private class CounterIterable(val max: Int) extends js.Iterable[Int] {
+    @JSName(js.Symbol.iterator)
+    def jsIterator(): js.Iterator[Int] = new CounterIterator(max)
+  }
+
+  @inline
```
Comment (sjrd) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/IterableTest.scala]: One more useless `@inline`

File: test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/IterableTest.scala [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+        }
+      }
+    }
+  }
+}
+
+class IterableTest {
+  import IterableTest._
+
+  def assertValue[T](expected: T, entry: js.Iterator.Entry[T]) = {
```
Comment (sjrd) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/IterableTest.scala]: Explicit result type missing.

**Comment by gzm0** [comment]
Updated.

**Comment by gzm0** [comment]
Jenkins, retest this please

**Comment by gzm0** [comment]
Note: The conversion from `js.Iterable` to `sc.Iterable` is potentially unsafe. MSN states "A generator object is both, iterator and iterable”. The `[Symbol.iterator]` method of a generator just returns itself, implying that it is OK (required?) for a `js.Iterable` to not return a fresh iterator. I am concerned that this might break some methods of `sc.Iterator`.

Maybe we should only allow the `toIterator` conversion and maybe add a utility to get a Scala iterator from a `js.Iterable`?

WDYT?

**Comment by sjrd** [comment]
The main goal of the conversion to `sc.Iterable` is to be able to use for comprehensions on such things, which is necessary to be on par with JavaScript and its for..of loop. Generators return themselves because they're essentially `Iterator`s to begin with, but JavaScript needs them to be `Iterable` to be usable in for..of. We don't need to present them as iterables, because we can use our for comprehensions directly on iterators too. 

So basically IMO what you describe is a non-issue for Scala.js, and we should not sacrifice the conversion to `sc.Iterable` for that.

**Comment by gzm0** [comment]
Ok, fair enough. So we'll just not introduce something like `Generatable`. Do we need to clarify the contract on `js.Iterable`? It might be just confusing...
